rfd/0094-kubernetes-node-joining.md

```diff
@@ -0,0 +1,335 @@
+field. The join logic would be similar to what is done with GitHub joining.
+
+The token spec could be extended in a second phase to support trusting other
+Kubernetes apiservers than the one where the auth is running. This would imply
+adding at least the following fields fields to the token:
+
+- Kubernetes cluster CA cert
+- Kubernetes Client certificate
+- Kubernetes Client key
+- Kubernetes user
+- Kubernetes server
+
+#### Security considerations
+
+Bound tokens mounted through projected volumes offer a solid level of security
+as those tokens are short-lived, automatically renewed and are revoked on the
+pod deletion. Those tokens are the pod's identity from a Kubernetes point of
+view.
+
+Legacy tokens offer less security guarantees as they are long-lived and not
+automatically rotated. Joining using them is still an improvement over static
+tokens because the Kubernetes-issued token doesn't need to leave the cluster
+by design: human operators and the CI/CD system should never be in contact with
+the token. Legacy tokens are mostly a user experience improvement rather than a
+security improvement.
+
+The Teleport Helm charts should always favor using bound tokens. Kubernetes
+support for bound tokens must be detected at the deploy time and the chart must
+adapt its behaviour to maximize security based on Kubernetes capabilities.
+
+To mitigate token stealing or replaying MiTM attack, the joining
+Teleport nodes should be configured with the ca_pins, as for regular tokens.
+
+Introducing the feature does not add new critical components that could cause
+availability issues. Creating new pods already requires a healthy apiserver.
+If the apiserver were to fail, this would not affect already joined proxies.
+
+#### Implications of adding the feature
+
+While joining Teleport proxy nodes is the main motivation of this proposal,
+other Teleport nodes would also benefit from this additional join method.
+
+Deploying nodes in the same Kubernetes cluster will become easier. This can be
+used to join all kinds of Teleport service (kubernetes, app, db, windows
+desktop, discovery), but also to join `tbot` and offer an easier Machine ID
+integration in Kubernetes. This would pave the way for a better experience for
+users hosting access plugins next to the Teleport control plane.
+
+```mermaid
+graph LR
+    subgraph cluster helm chart
+    cluster
+    tBot
+
+    subgraph accessPlugin chart
+    accessPlugin
+    end
+
+    cluster --- |joined using the 'kube' method| tBot
+    tBot --> |store certs| secret
+    secret --- |mounted| accessPlugin
+
+    accessPlugin --> |connects with its MachineID certs| cluster
+    end
+```
+
+Joining nodes from outside the Kubernetes cluster would imply generating and
+exporting a Kubernetes token from the service account. While this pattern is
+feasible, it does not provide much added value compared to generating a static
+token from Teleport directly.
```

Comment:
Would it make sense to provide more details on how it would work, maybe in a "future work" section? I just feel like having a "Kubernetes" join method and not supporting agents joining from other Kube clusters is strange. Like, if you run everything on Kubernetes, then you'd probably want to use the same join method even if your stuff is in different clusters? For comparison, IAM join method for example supports nodes joining from different AWS accounts.

Reply:
It should work the same way, except Teleport uses the kubeconfig from the resource instead of the `inCluster` config.

My biggest concern is that managed kubernetes hostings like GKE, EKS require more than a kubeconfig to interact with them. Suddenly we would have to support all the ways to pass service accounts to Teleport + support each cloud's auth provider in the go code. The spec would look like:

```yaml
kind: token
version: v2
metadata:
  name: proxy-token
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: ["Proxy"]
  k8s:
    allow:
    - service_account: "my-namespace:my-service-account"
      authType: "kubeconfig|gcp|aws|..."
      kubeconfig:
        certificate-authority-data: base64
        server: http://1.2.3.4
        # either set token or user & client-*
        token: my-secret-token
        user: client-sa-name
        client-certificate-data: base64
        client-key-data: base64
      gcp:
        # optional SA account key/name because you can use ambiant credentials or workload identity
        service-account-key: base64
        service-account-name: "foo@sa-project.iam.gserviceaccount.com"
        # mandatory
        project: cluster-project
        location: us-east1
        cluster-name: my-cluster
      aws:
        # optional SA account id/secret because you can use ambiant credentials or aws-iam-controller
        key-id: ASDFGHJK
        key-secret: QWERTYUIOP
        # mandatory
        region: us-east-1
        account-id: 1234567890
        cluster-name: my-cluster
      azure:
        # [...]
```

The users would have to map the cloud's service account to a kubernetes user/group and create a rolebinding to allow this group to use the `TokenReview` API. This step is cloud-specific and can be straightforward (as in GKE) or complex (hello AWS).

I'll add those details to the RFD so they can be reviewed at the same time by the auditors.

Reply:
done